src/main/java/net/wimpi/modbus/io/ModbusTCPTransport.java

```diff
@@ -95,16 +97,30 @@ public void close() throws IOException {
     @Override
     public void writeMessage(ModbusMessage msg) throws ModbusIOException {
         try {
-            msg.writeTo(m_Output);
+            byte[] serializedMessage = serialize(msg);
+            m_Output.write(serializedMessage);
             m_Output.flush();
             // write more sophisticated exception handling
-        } catch (Exception ex) {
+        } catch (IOException ex) {
             final String errMsg = "I/O exception - failed to write";
             logger.error(errMsg, ex);
             throw new ModbusIOException(String.format("%s: %s", errMsg, ex.getMessage()));
         }
     }// write
 
+    private byte[] serialize(ModbusMessage msg) throws IOException {
+        try (
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
```

Comment:
I would recommend to size the byte array according to `msg.getOutputLength()`.

Reply:
Note that it will not be exact either (the actual serialized message here **seems** to be 2 bytes shorter). But yes, that's a more optimal estimate than the default 32 bytes.